.grit/patterns/jest_to_vitest.md

```diff
@@ -0,0 +1,149 @@
+---
+title: Jest to Vitest
+---
+
+Convert Jest test to Vitest.
+
+tags: #migration, #js
+
+```grit
+engine marzano(0.1)
+language js
+
+pattern updated_imports() {
+  call_expression($function) where {
+    $function <: or {
+      js"test",
+      js"expect",
+      js"describe",
+      js"it",
+      js"beforeEach",
+      js"afterEach",
+      js"beforeAll",
+      js"afterAll"
+    } where {
+      $source = `'vitest'`,
+      $function <: ensure_import_from($source),
+      $vi = `vi`,
+      $vi <: ensure_import_from($source)
+    }
+  }
+}
+
+or {
+  // Imports
+  updated_imports(),
+  `import $_ from 'jest'` => .,
+
+  `jest.requireActual($module)` => `await vi.importActual($module)`,
+
+  `JEST_WORKER_ID` => `VITEST_POOL_ID`,
+
+  // module mocks
+  `jest.mock($module, $mockImplementation)` where { 
+    $mockImplementation <: arrow_function($body) where { 
+      $body <: literal_value() => `({ 
+        default: $body 
+      })`
+    }
+  } => `vi.mock($module, $mockImplementation)`,
+  `jest.mock` => `vi.mock`,
+
+  // done callback
+  `$sym($description, $callback)` where {
+    and {
+        $sym <: contains {
+            or { `it`, `test`}
+        },
+        $callback <: arrow_function($body, $parameters, $parameter) where {
+            or {
+                // For matching `done => { ... }`
+                ! $parameter <: .,
+                // For matching `(done) => { ... }`
+                $parameters <: contains r"\w+",
```

Comment:
What are you attempting to accomplish with this? Regex should not be necessary.

Reply:
Actually this can catch the cases where the devs don't use `done` term as the first arg, maybe we can update pattern to `.+` or Could you please suggest a condition for checking non-empty `$parameters`?

Reply:
Thanks @morgante for the suggestions, but when I tried it in [this place](https://github.com/getgrit/js/pull/208/files#diff-69db737d7970297c49ae2a0e9294dba9fb686c4b712b424bed3e390c96a6f40eR87) I doesn't work as expected (tests are failed), not sure what I'm doing wrong.

Could you please take a look

Reply:
Your link just points to another regex. Do *NOT* use a regex for this.

I am confident the approach I described works. Here's a sample link: https://app.grit.io/studio?key=x3FwwxrD607qwH-pmnvMh

Reply:
The link in my comment is to show you the position where I used your suggestion but it DIDN'T WORK.
Maybe it's from `$parameters` from prepared `function` and `arrow_function` patterns, I'm going to check...

Thanks for the sample link 👍

Reply:
The link in my comment is to show you the position where I used your suggestion but it DIDN'T WORK.
Maybe it's from `$parameters` from prepared `function` and `arrow_function` patterns, I'm going to check...

Thanks for the sample link 👍